flink-streaming-java/src/main/java/org/apache/flink/streaming/api/operators/BackendRestorerProcedure.java

```diff
@@ -141,14 +141,17 @@ public T createAndRestore(@Nonnull List<? extends Collection<S>> restoreOptions)
 	private T attemptCreateAndRestore(Collection<S> restoreState) throws Exception {
 
 		// create a new, empty backend.
-		final T backendInstance = instanceSupplier.get();
+		final T backendInstance = instanceSupplier.apply(restoreState);
 
 		try {
 			// register the backend with the registry to participate in task lifecycle w.r.t. cancellation.
 			backendCloseableRegistry.registerCloseable(backendInstance);
 
 			// attempt to restore from snapshot (or null if no state was checkpointed).
-			backendInstance.restore(restoreState);
+			// TODO we could remove this invocation when moving all backend's restore into builder
+			if (!backendInstance.getClass().getName().contains("RocksDBKeyedStateBackend")) {
```

Comment:
I also think this is not clean actually. However, since the `restore` method of `RocksDBKeyedStateBackend` is still required by the `restoreSnapshot` method in `StateBackendTestContext`, and if we make it does nothing `RocksDBTtlStateTest` will fail, I chose a kinda "ugly" way...

Reply:
I assumed already that the problem is with the tests, but is it hard to solve? For example, it seems like in `StateBackendTestContext`, we could simply merge the create and restore functions. They are always called together anyways, so if we use the builder here, that should fix it?

Reply:
For `StateBackendTestContext.restoreSnapshot` it reuses the pre-created keyed backend instance to do restores, so I guess you mean to merge the create and restore functions in `TtlStateTestBase`? It could work for rocksdb backend but not for others like heap backend, since the builder pattern is still not applied there. I believe this could be resolved after the refactoring of all backends done and totally agree to create a JIRA before this one is merged. But for the issue here, you prefer me to also apply a class-name checker in `StateBackendTestContext` or leave it as is? Thanks.

Reply:
Yes, I speak of a merge between create and restore. I think until the refactoring of all backends to builder pattern is complete you can still use call `restore`on the keyed state backend instances. But once we changed this for RocksDB, if I see correctly, we can also let RocksDB have an empty restore method and then it would work again.

Reply:
Yes you are correct, I've already checked and confirmed the merge of create and restore could work well for rocksdb. Back to the review here, is it ok to keep this `contains` logic for the time being and fix it later in the to-be-created JIRA for other backend's refactoring? Thanks.

Reply:
I think that when the tests we discussed are changed then we should be able to make `RocksDBKeyedStateBackend::restore(...)` an empty method, because I think keeping the method was only required as workaround for `RocksDBStateBackendTest`. `RocksDBStateBackendTest` can be refactored to avoid this, as described in one of my other comments. Once `RocksDBKeyedStateBackend::restore(...)` is an empty method, you can remove the extra check and just always call restore. The same for all the tests where you currently added `if (backend instanceof HeapKeyedStateBackend) {backend.restore(...); }`.

Reply:
`RocksDBTtlStateTest` and `RocksDBStateBackendMigrationTest` also counts, but yes I've found the way to fully empty `restore` method in `RocksDBKeyedStateBackend`, will have the fix in the next commit.